pallets/reward-distribution/src/lib.rs

```diff
@@ -99,5 +156,188 @@ pub mod pallet {
+					return amount.transfer(&Self::fee_pool_account_id(), &beneficiary)
+				},
+				Some(remaining) => {
+					// Use the available funds from the fee pool
+					let available_amount: currency::Amount<T> =
+						Amount::new(available_native_funds, reward_currency_id);
+					available_amount.transfer(&Self::fee_pool_account_id(), &beneficiary)?;
+					// Mint the rest
+					T::Balances::deposit_creating(&beneficiary, remaining);
+				},
+			}
+			Ok(())
+		} else {
+			//we need no checking of available funds, since the fee pool will ALWAYS have enough
+			// collected fees of the wrapped currencies
+			let amount: currency::Amount<T> = Amount::new(reward, reward_currency_id);
+			amount.transfer(&Self::fee_pool_account_id(), &beneficiary)
+		}
+	}
+
+	pub fn execute_on_init(height: T::BlockNumber) {
+		//get reward per block
+		let reward_per_block = match RewardPerBlock::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("Reward per block is None");
+				return
+			},
+		};
+
+		if let Err(_) = ext::security::ensure_parachain_status_running::<T>() {
+			return
+		}
+
+		let rewards_adapted_at = match RewardsAdaptedAt::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("RewardsAdaptedAt is None");
+				return
+			},
+		};
+
+		let mut reward_this_block = reward_per_block;
+
+		//update the reward per block if decay interval passed
+		if let Ok(expired) = ext::security::parachain_block_expired::<T>(
+			rewards_adapted_at,
+			T::DecayInterval::get() - T::BlockNumber::one(),
+		) {
+			if expired {
+				let decay_rate = T::DecayRate::get();
+				reward_this_block = (Perquintill::one() - decay_rate).mul_floor(reward_per_block);
+				RewardPerBlock::<T>::set(Some(reward_this_block));
+				RewardsAdaptedAt::<T>::set(Some(height));
+			}
+		} else {
+			log::warn!("Failed to check if the parachain block expired");
+		}
+
+		//TODO how to handle error if on init cannot fail?
+		let _ = Self::distribute_rewards(reward_this_block, T::GetNativeCurrencyId::get());
+	}
+
+	//fetch total stake (all), and calulate total usd stake in percentage across pools
+	//distribute the reward accoridigly
+	fn distribute_rewards(
+		reward_amount: BalanceOf<T>,
+		reward_currency: CurrencyId<T>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		//calculate the total stake across all collateral pools in USD
+		let total_stakes = ext::pooled_rewards::get_total_stake_all_pools::<T>()?;
+
+		let mut total_stake_in_usd = BalanceOf::<T>::default();
+		for (currency_id, stake) in total_stakes.clone().into_iter() {
+			let stake_in_usd = T::OracleApi::currency_to_usd(&stake, &currency_id)?;
+			total_stake_in_usd = total_stake_in_usd.checked_add(&stake_in_usd).unwrap();
+		}
+		//distribute the rewards to each collateral pool
+		let mut error_reward_accum = BalanceOf::<T>::zero();
+		for (currency_id, stake) in total_stakes.into_iter() {
+			let stake_in_usd = T::OracleApi::currency_to_usd(&stake, &currency_id)?;
```

Comment:
I think I understand, we can save into an intermediate vector of `stake_in_usd` to avoid recalculating the price in USD in the first line of the second iteration.

Reply:
Exactly!